lib/appFactory.js

```diff
@@ -0,0 +1,85 @@
+const express = require('express');
+const path = require('path');
+const { loggingMiddleware } = require('./logger');
+const indexRouter = require('./routes/index');
+const signupRouter = require('./routes/signup');
+const webFingerRouter = require('./routes/webfinger');
+const oAuthRouter = require('./routes/oauth');
+const storageCommon = require('./routes/storage_common');
+const errorPage = require('./util/errorPage');
+const helmet = require('helmet');
+const shorten = require('./util/shorten');
+
+module.exports = function ({ hostIdentity, jwtSecret, account, store, basePath = '' }) {
+  if (basePath && !basePath.startsWith('/')) { basePath = '/' + basePath; }
+
+  const app = express();
+  app.locals.basePath = basePath;
+
+  // view engine setup
+  app.engine('.html', require('ejs').__express);
+  app.engine('.xml', require('ejs').__express);
+  app.set('view engine', 'html');
+  app.set('views', path.join(__dirname, 'views'));
+
+  express.static.mime.define({ 'text/javascript': ['js'] });
+
+  app.set('account', account);
+
+  app.use(loggingMiddleware);
+  app.use(helmet({
+    contentSecurityPolicy: {
+      directives: {
+        sandbox: ['allow-scripts', 'allow-forms', 'allow-popups', 'allow-same-origin'],
+        defaultSrc: ['\'self\''],
+        scriptSrc: ['\'self\''],
+        scriptSrcAttr: ['\'none\''],
+        styleSrc: ['\'self\''],
+        imgSrc: ['\'self\''],
+        fontSrc: ['\'self\''],
+        objectSrc: ['\'none\''],
+        childSrc: ['\'none\''],
+        connectSrc: ['\'none\''],
+        baseUri: ['\'self\''],
+        frameAncestors: ['\'none\''],
+        formAction: (process.env.NODE_ENV === 'production' ? ['https:'] : ['https:', 'http:']), // allows redirect to any RS app
+        upgradeInsecureRequests: []
+      }
+    }
+  }));
+  app.use(express.urlencoded({ extended: false }));
+  app.use(`${basePath}/assets`, express.static(path.join(__dirname, 'assets')));
+
+  app.use(`${basePath}/`, indexRouter);
+
+  app.use(`${basePath}/signup`, signupRouter);
+
+  app.use([`${basePath}/.well-known`, `${basePath}/webfinger`], webFingerRouter);
+
+  app.use(`${basePath}/oauth`, oAuthRouter(hostIdentity, jwtSecret));
```

Comment:
It's interesting your choice of using the routing middleware implementation in express.  As a matter of preference, I like keeping all of my routing specifications in one file and then delegate the functional implementations to external "Controller" files for each endpoint being declared.  Only when the overall routing structure gets too cumbersome do I shunt sections to routing middleware.

I see no issues with how you implemented this.  I just wanted to provide a difference of opinion in my approach versus yours.

Reply:
The main reason is sharing helper functions — for example, in `routes/oauth.js`, there's a helper function `validOAuthRequest` that is used by both routes:

```javascript
  router.get('/:username',
    redirectToSSL,
    formOrQueryData,
    sanityCheckUsername,
    validOAuthRequest,
    function (req, res) {
        // ...
    });

  router.post('/',
    secureRequest,
    formOrQueryData,
    sanityCheckUsername,
    validOAuthRequest,
    async function (req, res) {
        // ...
    }
  );

  function validOAuthRequest (req, res, next) {
        // ...
    next();
  }


```